### PR TITLE
chore(test): Use SWC as Jest transformer

### PR DESCRIPTION
**Pull request by Thisen**
Use [@swc/jest](https://github.com/swc-project/jest) as Jest transformer, for faster test runs.

Locally I saw this:
- Before: ~10 secs
- With @swc/jest: ~6.6 secs

That's about 41% faster.

This is not much we gain currently, but as discussed between us maintainers, we want to focus more on tests, so this change will be better for us, as time goes.

NOTE: This is an experimental and a suggestion, I'm fine with this landing. 😄

### CODE REVIEW

File: tests/async.test.tsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { StrictMode, Suspense, useEffect, useRef } from 'react'
+import React, { StrictMode, Suspense, useEffect, useRef } from 'react'
```
Comment (dai-shi) [inline tests/async.test.tsx:1]: It's our intention to prefer jsx-transform. If swc doesn't support it, can we somehow hack it without changing the code please?

File: tests/async.test.tsx [hunk 2]
```diff
@@ -498,7 +498,7 @@ it('set promise atom value on write (#304)', async () => {
     set(
       countAtom,
       Promise.resolve(get(countAtom)).then(
-        (c) => new Promise((r) => setTimeout(() => r(c + 1), 500))
+        (c) => new Promise((r) => setTimeout(() => r(c + 1), 1))
```
Comment (dai-shi) [inline tests/async.test.tsx:501]: not sure how this works. do they use the fake timer by default?

File: tests/async.test.tsx [hunk 3]
```diff
@@ -767,9 +767,7 @@ it('non suspense async write self atom with setTimeout (#389)', async () => {
 })
 
 it('should override promise as atom value (#430)', async () => {
-  const countAtom = atom(
-    new Promise<number>((r) => setTimeout(() => r(-1), 3600 * 1000))
```
Comment (dai-shi) [inline tests/async.test.tsx:771]: this means the promise should never resolve. sorry for the confusion.
let's make it `new Promise<number>)(() => {}) // infinite promise`.

**Review by dai-shi — CHANGES_REQUESTED** [review]
we also need to make sure that all CI tests pass, fairly stably.

**Comment by aulneau** [comment]
I use this in a number of my own projects, great add!

File: tests/async.test.tsx (hunk 2 above)
Comment (Thisen) [reply tests/async.test.tsx]: Can you elaborate?

File: tests/async.test.tsx (hunk 3 above)
Comment (Thisen) [reply tests/async.test.tsx:771]: Fixed :)

File: tests/async.test.tsx (hunk 2 above)
Comment (dai-shi) [reply tests/async.test.tsx]: In my experience with the previous one, this doesn't pass CI.
So, we would like to a) see if it passes, and b) know why.

Comment (Thisen) [reply tests/async.test.tsx]: You need to it forward then :)

**Review by dai-shi — COMMENTED** [review]
Alright, the code looks good.
We make this type of PRs for zustand and valtio too, before merging.
Would you be able to work on those two, or do you want me or someone else to work on them?

**Comment by Thisen** [comment]
> Alright, the code looks good. We make this type of PRs for zustand and valtio too, before merging. Would you be able to work on those two, or do you want me or someone else to work on them?

Happy to do it for Valtio and Zustand aswell.

File: .github/workflows/test-multiple-versions.yml [hunk 4]
```diff
@@ -65,6 +65,7 @@ jobs:
         run: |
           sed -i~ '1s/^/import React from "react";/' tests/*.tsx tests/*/*.tsx
           sed -i~ 's/automatic/classic/' babel.config.js
+          sed -i~ 's/automatic/classic/' .swcrc
```
Comment (dai-shi) [inline .github/workflows/test-multiple-versions.yml:68]: 👏

**Review by dai-shi — APPROVED** [review]

**Comment by dai-shi** [comment]
@Thisen I think it's because of this PR, but now jest error reporting doesn't show the correct error location (line number) in local development. This is pretty annoying. Hope you find a solution.

**Comment by Thisen** [comment]
> @Thisen I think it's because of this PR, but now jest error reporting doesn't show the correct error location (line number) in local development. This is pretty annoying. Hope you find a solution.

I'll look into it.

File: tests/async.test.tsx [hunk 5]
```diff
@@ -27,7 +27,7 @@ it('does not show async stale result', async () => {
     const [count, setCount] = useAtom(countAtom)
     const onClick = async () => {
       setCount((c) => c + 1)
-      await new Promise((r) => setTimeout(r, 10))
+      await new Promise((r) => setTimeout(r, 1))
```
Comment (dai-shi) [inline tests/async.test.tsx:30]: @Thisen Unfortunately, this is causing the CI failure several times, so let me revert it.

**Pull request merged**